### PR TITLE
[WIP] Mongodb support

### PR DESCRIPTION
**Pull request by Zayon**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #461 #1248
| License       | MIT
| Doc PR        | none

Filters need a rework because `FilterInterface` use `Doctrine\ORM\QueryBuilder` which is not compatible with mongodb (and other potential providers)

I removed fos_user from behat test for mongo db as didn't manage to understand the problem.

@dunglas @Simperfit @samvdb @ihr-it-projekt

### CODE REVIEW

File: src/Bridge/Doctrine/Filter/AbstractFilter.php [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Bridge\Doctrine\Filter;
```
Comment (soyuka) [inline src/Bridge/Doctrine/Filter/AbstractFilter.php]: Why is this file part of this PR?

File: src/Bridge/Doctrine/Filter/FilterInterface.php [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+use ApiPlatform\Core\Api\FilterInterface as BaseFilterInterface;
+use ApiPlatform\Core\Bridge\Doctrine\Orm\Util\QueryNameGeneratorInterface;
+use Doctrine\ORM\QueryBuilder;
+
+/**
+ * Doctrine ORM filter interface.
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+interface FilterInterface extends BaseFilterInterface
```
Comment (soyuka) [inline src/Bridge/Doctrine/Filter/FilterInterface.php]: Same, should not be part of this PR

File: src/Bridge/Doctrine/Filter/AbstractFilter.php (hunk 1 above)
Comment (Zayon) [inline src/Bridge/Doctrine/Filter/AbstractFilter.php]: It was part of this one : #461 and I didn't removed it for now.

File: src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php [hunk 3]
```diff
@@ -64,11 +64,12 @@ public function getConfigTreeBuilder()
                         ->booleanNode('force_eager')->defaultTrue()->info('Force join on every relation. If disabled, it will only join relations having the EAGER fetch mode.')->end()
                     ->end()
                 ->end()
+                ->booleanNode('enable_doctrine_mongodb_odm')->defaultValue(false)->info('Enable Doctrine MongoDB ODM integration.')->end()
+                ->booleanNode('enable_doctrine_orm')->defaultValue(true)->info('Enable Doctrine ORM integration.')->end()
```
Comment (soyuka) [inline src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php]: prefer:

```
doctrine:
  mongodb_odm: false
  doctrine_orm: true
```

Comment (Zayon) [inline src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php]: do you mean the format yml over php ?

File: src/Bridge/Doctrine/Filter/AbstractFilter.php (hunk 1 above)
Comment (soyuka) [inline src/Bridge/Doctrine/Filter/AbstractFilter.php]: I was surprised because there is no diff. But yeah, we moved this in `ApiPlatform\Core\Bridge\Doctrine\Orm\Filter` since this is why there are no diffs :p. 
My bad!

File: src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php (hunk 3 above)
Comment (soyuka) [inline src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php]: Sorry, this would be the end yaml configuration. I'd put these two values in a sub `arrayNode` (`doctrine`) here:

```
->arrayNode('doctrine')
                    ->addDefaultsIfNotSet()
                    ->children()
```

**Comment by Zayon** [comment]
Coveralls seems to be failing because it doesn't test the mongodb case. I don't know how to configure that, should I create a .coveralls file ?

Also travis is failing only on php 7.0 and the stack trace is weird, the error comes from doctrine vendor.

I did the best I can but I need some help to finish this PR.

@soyuka

**Comment by soyuka** [comment]
I think this has been fixed in https://github.com/doctrine/mongodb-odm/pull/1543 (`@1.1.3`).

In `composer.json`, the `mongodb-odm` version is fixed at:

```
        "doctrine/mongodb-odm": "1.1.0", 
``` 

Use:
```
      "doctrine/mongodb-odm": "^1.1.5", 
```

**Comment by Zayon** [comment]
`"doctrine/mongodb-odm": "^1.1.5"` did not work :disappointed:

**Comment by soyuka** [comment]
Hmm yes you're right, not sure why it uses `doctrine/annotations` v1.5, it's using `v1.4` with php7 on other pull requests. Only thing I see is the `ignore-platform-reqs` that could be the root cause.

**Comment by Zayon** [comment]
Travis is Green ! :tada: 

I removed `ignore-platform-reqs` from composer calls but I add to add mongodb dependencies as provide in `composer.json`. I think you don't want this as part of api-platform, how should we provide mongodb dependencies to travis ?

File: composer.json [hunk 4]
```diff
@@ -34,6 +34,9 @@
         "behatch/contexts": "^2.6",
         "doctrine/doctrine-bundle": "^1.6.3",
         "doctrine/orm": "^2.5.2",
+        "doctrine/mongodb-odm": "^1.1.5",
+        "doctrine/mongodb-odm-bundle": "~3.0",
+        "ext-mongo": "*",
```
Comment (soyuka) [inline composer.json]: `ext-mongodb` here should be sufficient.

Comment (Zayon) [inline composer.json]: It's not, I tested.

Comment (Zayon) [inline composer.json]: https://travis-ci.org/api-platform/core/builds/260093026

Comment (soyuka) [inline composer.json]: Indeed, you need to [enable the extension in travis](https://docs.travis-ci.com/user/languages/php#PHP-extensions).

For example, add : 
```
before_install: 
 - echo "extension = mongodb.so" >> ~/.phpenv/versions/$(phpenv version-name)/etc/php.ini
```

also, it's `ext-mongodb` for php 7 not `ext-mongo`

Comment (Zayon) [inline composer.json]: https://travis-ci.org/api-platform/core/jobs/260125998

> doctrine/mongodb 1.5.0 requires ext-mongo ^1.5 -> the requested PHP extension mongo is missing from your system.

It seem we need both `ext-mongodb` and `ext-mongo` for some reason

Comment (soyuka) [inline composer.json]: Weird, try to enable both but it should require only `ext-mongodb` for php7+ (see linked travis docs).

**Comment by Simperfit** [comment]
Thanks @Zayon for finishing this. We should test against a demo application using mango to be sure that this is working properly, I didn't had the time to do so unfortunately.

**Comment by Zayon** [comment]
Hi @Simperfit ,

With my team we are going to build an app with api platform and mongodb, so we are going to test this code and push patches if needed.

Also, I can't find a solution with that https://travis-ci.org/api-platform/core/jobs/260129477
`doctrine/mongodb` require `ext-mongo` which is php <= 5.99.99 only.

**Comment by soyuka** [comment]
No don't use pecl. Can you try this:

```
 - echo "extension = mongodb.so" >> ~/.phpenv/versions/$(phpenv version-name)/etc/php.ini
 - echo "extension = mongo.so" >> ~/.phpenv/versions/$(phpenv version-name)/etc/php.ini
```

Don't mind the warnings.

**Comment by Zayon** [comment]
I added 

```
"provide": {
    "ext-mongo": "^1.6.14"
}
```

as stated here : http://docs.doctrine-project.org/projects/doctrine-mongodb-odm/en/latest/reference/introduction.html#using-php-7

> You can use Doctrine MongoDB ODM with PHP 7, but there are a few extra steps during the installation. Since the legacy driver (referred to as ext-mongo) is not available on PHP 7, you will need the new driver (ext-mongodb) installed and use a polyfill to provide the API of the legacy driver.
>
> To do this, you have to require alcaeus/mongo-php-adapter before adding a composer dependency to ODM. To do this, run the following command:
>
> `composer require "alcaeus/mongo-php-adapter"`
>
> Next, manually add a provide section to your composer.json:
>
> ```
> "provide": {
>    "ext-mongo": "1.6.14"
>}
>```
>
>This section needs to be added to work around a composer issue with libraries providing platform packages (such as ext-mongo). Now, you may install ODM as described above:
>
> `composer require "doctrine/mongodb-odm"`

**Comment by soyuka** [comment]
haha nice one :+1: !

So, if I understand correctly, some behat tests are actually running successfully on mongodb? Last thing we need to cover the full behat suite is to add filters right?

> We should test against a demo application using mango to be sure that this is working properly, I didn't had the time to do so unfortunately.

@Simperfit not sure how this is relevant.

Next steps are:
- code review
- more code review
- filters implementation
- docs

To ease the git management of this, WDYT about having a `mongodb` branch on `api-platform/core` that we could keep up to date with master while it's being built? Just saying this because this may not end up in `3.0`.

**Comment by Zayon** [comment]
It also misses sub-resources. I don't know if this is relevant with mongodb.

Also, fos user as it use doctrine orm.

**Comment by Simperfit** [comment]
I think this is relevant because we need to be able to be 100% sure that
this is working with an app.

Ill do a review asap ;)

Le mer. 2 août 2017 à 15:17, Antoine Bluchet <notifications@github.com> a
écrit :

**Comment by ad3n** [comment]
@Zayon : It's mean we can use orm and odm both? Or only use one between orm and odm?

**Comment by Zayon** [comment]
I did not try to use both. In theory it might work with a bit of tweaking. There will be limitations I think. For example I don't think it will possible to mix 2 endpoints from 2 data sources. 
If you really want to use 2 data sources you should consider not using doctrine?

**Comment by ad3n** [comment]
@Zayon : when you can add extra config in `@ApiResource` to handle data source, example `@ApiResource(driver="odm")`, may we can use both

**Comment by ad3n** [comment]
If we can do that, we can implement `http://docs.doctrine-project.org/projects/doctrine-mongodb-odm/en/latest/cookbook/blending-orm-and-mongodb-odm.html` this feature...

**Comment by dunglas** [comment]
Any news on this one? Would be awesome to be able to merge it!

**Comment by Zayon** [comment]
Hi, we are no longer working on this as our project as been postponed.

@aldelplace should have the lastest info on this.

**Comment by soyuka** [comment]
@dunglas AFAIR we successfully made tests pass on this one!

In my latest comment, what's needed to get this merged is:

- code review
- more code review
- filters implementation
- docs

**Comment by adelplace** [comment]
Hi,

@dunglas to me there still lot of work. Actually the behat tests weren't well wired to work with doctrine odm. We need to create documents to "translate" the current testing entities and adapt some tests to fit with mongo limitations which is a heavy work and I don't talk about the filters implementation...

Unfortunately, as our project using this api is stopped for now I can't give myself enough time to finish this work.

**Comment by samvdb** [comment]
The current filtering logic implemented in this branch should be reworked before merging into master.

**Comment by guizmo51** [comment]
hey, I would like to help you. I forked this repo but I can't retrieve the specific branch for mongodb.

**Comment by Zayon** [comment]
Hi, you can do `git fetch origin pull/1293/head:mongodb` to create a "mongodb" branch based on this PR.
But you won't be able to update the PR directly, you'll have to either create your own PR or create a PR on my fork https://github.com/zayon/core

Maybe someone from the api-platform team should create a dedicated branch a this repo @dunglas @Simperfit @soyuka ?

@adelplace should push his last work on this PR.

**Comment by adelplace** [comment]
My little contribution is now on this branch ;)

**Comment by giovannialbero1992** [comment]
This PR is abandoned? What can we do to complete this enhancement? Ty

**Comment by soyuka** [comment]
@giovannialbero1992 : 

> 
rebase against master
code review
more code review
filters implementation
docs

**Comment by ad3n** [comment]
any update?

**Comment by guizmo51** [comment]
@ad3n unfortunately i can't help for this project but it's open source, you cannot come here and just ask "any update...?" and the previous message (written 5 days ago) is clear, @soyuka listed tasks to do before. 

I'm also interested in this features but two options if I want to see this PR released  :  
- be quiet, wait the work of others dev
or
- help by reading previous messages and by trying to work on tasks listed
